### ISSUE TITLE
No 4x downsampling processing: Serious Error happens in th  Inference code!!!

### ISSUE THREAD

**Issue by KindleHe** [ppwwyyxx/tensorpack#706]
The code below is executed after run `python enet-pat.py --apply --load /path/to/model --lowres input.png` command. However, there is **no 4x downsampling processing** for the original input images. According to [your released result ](https://github.com/ppwwyyxx/tensorpack/files/1522218/results.tar.gz), the resolution of your released images is all `2048x2048`(no matter your re-implement or the paper's author's model), which is 4x times of the paper author's released original results  `512x512` in both height and width. This also reflects that your  [result image](https://github.com/ppwwyyxx/tensorpack/blob/master/examples/SuperResolution/enhancenet-demo.jpg) may be tested by inputting the "512x512" original HR images, instead of 4x downsampling images. 
To evaluate your released model's effects, I input the 4x downsampling image (see the small images below) to the function "apply" with your [model](http://models.tensorpack.com/SuperResolution/), get a far worse result (see the big size images below) than author's model or released results.

```
def apply(model_path, lowres_path="", output_path='.'):
    assert os.path.isfile(lowres_path)
    assert os.path.isdir(output_path)
    lr = cv2.imread(lowres_path).astype(np.float32)
    baseline = cv2.resize(lr, (0, 0), fx=4, fy=4, interpolation=cv2.INTER_LANCZOS4)
    LR_SIZE_H, LR_SIZE_W = lr.shape[:2]

    predict_func = OfflinePredictor(PredictConfig(
        model=Model(LR_SIZE_H, LR_SIZE_W),
        session_init=get_model_loader(model_path),
        input_names=['Ilr'],
        output_names=['prediction']))

    pred = predict_func(lr[None, ...])
    p = np.clip(pred[0][0, ...], 0, 255)

    cv2.imwrite(os.path.join(output_path, "predition.png"), p)
    cv2.imwrite(os.path.join(output_path, "baseline.png"), baseline)
```
![baseline-enet](https://user-images.githubusercontent.com/17778053/37782869-fdd8f466-2e2e-11e8-9486-7ef67e5c4796.png)
![predition-enet](https://user-images.githubusercontent.com/17778053/37782871-000442b8-2e2f-11e8-921c-d688023c3245.png)
![eagle-input](https://user-images.githubusercontent.com/17778053/37783707-1ef58b4e-2e31-11e8-9654-8728a36a4347.png)

![predition](https://user-images.githubusercontent.com/17778053/37782876-03388426-2e2f-11e8-98d9-af2460bd5fec.png)

**Comment by ppwwyyxx** [comment]
@PatWie

**Comment by KindleHe** [comment]
Another problem is that the accuracy of the discriminator keeps about 0.6~0.7 after 217 epochs on train2017.zip. The loss don't decrease for many epochs. Here is my output parameters:
```
Epoch 217 (global_step 601524) finished, time:10 minutes 27 seconds.
[0322 16:36:23 @saver.py:84] Model saved to train_log/enet-pat/model-601524.
[0322 16:36:23 @monitor.py:428] GAN_loss/discrim/accuracy: 0.61338
[0322 16:36:23 @monitor.py:428] GAN_loss/discrim/loss: 0.64238
[0322 16:36:23 @monitor.py:428] GAN_loss/gen/accuracy: 0.17531
[0322 16:36:23 @monitor.py:428] GAN_loss/gen/loss: 0.91403
[0322 16:36:23 @monitor.py:428] QueueInput/queue_size: 50
[0322 16:36:23 @monitor.py:428] additional_losses/loss_LA: 1.8281
[0322 16:36:23 @monitor.py:428] additional_losses/loss_LP1: 0.3446
[0322 16:36:23 @monitor.py:428] additional_losses/loss_LP2: 0.26865
[0322 16:36:23 @monitor.py:428] additional_losses/loss_LT1: 0.026572
[0322 16:36:23 @monitor.py:428] additional_losses/loss_LT2: 0.026721
[0322 16:36:23 @monitor.py:428] additional_losses/loss_LT3: 0.05257
```
I find your experiment records in the [link](https://github.com/ppwwyyxx/tensorpack/pull/527)(a copy is below), I have two doubts: 

**1)why 40 epoches achieves 0.99414 discrim/accuracy while my 217get only 0.61338?**    
**2)why the loss_PA between your (8.9464) and my (1.8281) is much different?**
**3) why your LT1 LT2 LT3 is so small whiel my is not?** 
**4) Could you tell me your training config and machine config?**
```
100%|#######################|19714/19714[59:02<00:00, 5.56it/s]          
[1206 04:49:54 @base.py:255] Epoch 40 (global_step 788560) finished, time:3542.78 sec.                                                             
[1206 04:49:55 @saver.py:82] Model saved to train_log/enet-pat/model-788560.                                                                       
[1206 04:49:55 @monitor.py:363] GAN_loss/discrim/accuracy: 0.99414       
[1206 04:49:55 @monitor.py:363] GAN_loss/discrim/loss: 0.019299          
[1206 04:49:55 @monitor.py:363] GAN_loss/gen/accuracy: 0.0026996         
[1206 04:49:55 @monitor.py:363] GAN_loss/gen/loss: 8.9464
[1206 04:49:55 @monitor.py:363] QueueInput/queue_size: 49.585
[1206 04:49:55 @monitor.py:363] add: 19.204
[1206 04:49:55 @monitor.py:363] additional_losses/loss_LA: 8.9464
[1206 04:49:55 @monitor.py:363] additional_losses/loss_LP1: 1.3098
[1206 04:49:55 @monitor.py:363] additional_losses/loss_LP2: 0.0015531
[1206 04:49:55 @monitor.py:363] additional_losses/loss_LT1: 2.3112e-37
[1206 04:49:55 @monitor.py:363] additional_losses/loss_LT2: 2.2718e-37
[1206 04:49:55 @monitor.py:363] additional_losses/loss_LT3: 2.3154e-37
[1206 04:49:55 @base.py:245] Start Epoch 41 ...
100%|#######################|19714/19714[59:03<00:00, 5.56it/s]
[1206 05:48:58 @base.py:255] Epoch 41 (global_step 808274) finished, time:3543.57 sec.
```

**Comment by KindleHe** [comment]
According to your reply in #541 , the below [website] (https://github.com/ppwwyyxx/tensorpack/pull/541) cannot open with remind "404 error", I am curious how you always obtain a model better than baseline, and what the baseline here is ?(the paper's model?)
```
Did the following in https://gist.github.com/ppwwyyxx/6da15f2d9087373635254d4e6fbe4891:
Used the paper's normalization
Fixed VGG scale
Used same learning rate for G and D
Used SeparateGANTrainer

And now I can always obtain a model better than baseline (in all the 3 times I've tried).
```

**Comment by ppwwyyxx** [comment]
For 4x: I think that's intended. 128->512 is superresolution. 512->2048 is also superresolution. The latter one is used.

The original code is [here](https://github.com/cgtuebingen/tensorflow-recipes/tree/master/EnhanceNet) (but it cannot run anymore because it should have copied GAN.py). The model there is still downloadable. But it's not compatible with code in tensorpack examples.

I changed the code a bit to make the results slightly better in my experiments, but I never succeeded in getting something as good as the demo image. So bad results you're getting is not a surprise to me.

**Comment by PatWie** [comment]
Honestly, it never came to my mind to _artificially_ downscale the input image _before_ running the inference in practice. Still, it makes no sense for me in this particular work with GANs (things are different for ENet-E and PSNR).

But if this is an issue for you @KindleHe, feel free to process the images with **this** implementation _without_ down-scaling. And add a down-scaling afterward on the final prediction when comparing to other implementations. When doing so you'll get

![out43](https://user-images.githubusercontent.com/6756603/37793736-c89eeae0-2e0f-11e8-8ee6-5f1cf967a1a7.png)
![out77](https://user-images.githubusercontent.com/6756603/37793739-ca4e41ba-2e0f-11e8-94a1-8f9ab0fec975.png)

(left to right: official EnhanceNet-PAT result, this implementation, ground-truth) on BSD100. 

To cite [Colorful Image Colorization](http://richzhang.github.io/colorization/)
> Welcome! Computer vision algorithms often work well on some images, but fail on others. Ours is like this too.

This is probably the case here as well. You will probably find images, where the author's implementation beats our version. I am ok with that as I do not care about this task.

One last thought: I am frequently wondering how people publish methods: Do we benchmark these algorithms against specific down-scaling implementations and on synthetic cases? Or do we want to solve real-world problems?
With/without anti-aliasing? Bicubic, bi-linear, Lanczos? OpenCV, PIL, TensorFlow implementation?
